Implement database reconnection with automatic retry logic

davidkopp opened the pull request:
Resolves #1296

ArneTR commented on tests/test_runner.py:
```diff
@@ -598,3 +600,105 @@ def test_print_logs_flag_with_iterations():
+    when database operations are likely occurring for metric storage.
+    """
+
+    def log_with_timestamp(message, prefix="TEST", start_time=None):
```
if this is a helper function and very general I think placing it in test_functions would be more expected

davidkopp replied:
Done.

ArneTR commented on tests/test_runner.py:
```diff
@@ -598,3 +600,105 @@ def test_print_logs_flag_with_iterations():
+        log_with_timestamp("Database should be available again after 3s wait", "DB RESTART", test_start_time)
+
+    # Start database restart in background thread
+    restart_thread = threading.Thread(target=restart_database)
```
I wonder if this is the nicest implementation. If the sole reason is that docker restart takes long to return I think a docker compose stop -t 0 and then a docker a docker start should do it. Since there should be no I/O also no thread is needed.

Would that help?

My issue with the thread: Everytime I see the Thread library used my brain switches to scanner mode to find the high complexity part of the code. So it makes code for me harder to read if there is no strong reason for it to use. Maybe also a me issue ...

ArneTR commented on lib/db.py:
```diff
@@ -1,13 +1,74 @@
 
+def with_db_retry(func):
+    @wraps(func)
+    def wrapper(self, *args, **kwargs):
```
The function looks good but I think it has a design issue.

If the connection is not on auto-commit you might be re-connection and retrying a part of the query only. This can be tricky problem to debug.

Example code :
```
conn.autocommit = False
query('INSERT 1 INTO table')
## Connection drops here
## connection gets reinstantiated
query('DELETE 1 FROM table') # only the delete query is now executed as first request was dropped
```

In order to find a solution here I believe either the queries need to be buffered ... or we could decide to still fail ... this however would defeat the purpose of this PR entierely i feel.

Currently unclear what the best implementation is without a clear view how complex a query buffering would be. Please investigate / make a concept.

2. Als the `copy()` statement is currently not profiting from the statement. This would need to be covered too.
To extend this question: Are there any other query methods that can be covered we should think of to future-proof?

3. How does the DB behave on keep-alive queries? Will they also lead to a reconnect?
To my understanding the postgres adapter does regular keep-alives. If they fail it will also complain. Can we hook into this?

davidkopp replied:
## Amendment to the 3. point "PostgreSQL keep-alive queries?"

psycopg3 includes a built-in `ConnectionPool` retry mechanisms. It is disabled by default, but can be enabled with the parameter `check`:

```python
self._pool = ConnectionPool(
    # ...
    check=ConnectionPool.check_connection
)
```

### Comparison

How does it compare to the custom `@with_db_retry` decorator implementation?

#### psycopg3 Built-in Limitations

**Measurement Interference:**
- `check=ConnectionPool.check_connection` executes `conn.execute("")` on every connection borrow
- Continuous background pool maintenance network calls
- Unnecessary network calls during measurement runs should be avoided

**Limited Error Scope:**
- No checks performed during actual query execution (`cur.execute()`)
- Cannot handle connections that fail mid-query
- Generic `PoolTimeout` exceptions with limited context

#### Custom `@with_db_retry` Advantages

**Measurement Integrity:**
- Zero network overhead during normal operation
- Only activates during actual database failures
- No background connection health checks

**Superior Error Handling:**
- Detects and retries failures during `cur.execute()`
- Classifies retryable vs non-retryable errors
- Complete pool recreation ensures clean state
- Detailed retry logging with error context

**Similar Retry Performance:**
Both mechanisms use comparable exponential backoff (1s, 2s, 4s, 8s...) with jitter.

### Conclusion

**Use the custom `@with_db_retry` mechanism** because:
- Preserves measurement accuracy by avoiding unnecessary network calls
- Better error detection scope (covers mid-query failures)
- Superior observability and debugging capabilities
- No compromise in retry performance

To be explicit that we don't want to use the check functionality of psycopg3, we should set it to `None`:

```python
self._pool = ConnectionPool(
    # ...
    check=None
)
```

davidkopp commented:
Thanks for the review. Changes made:
- added test cases for the db methods
- added the `with_db_retry` decorator to the `copy_from` method
- moved `log_with_timestamp` to `test_functions.py`

Compare:
https://github.com/green-coding-solutions/green-metrics-tool/compare/8bb0b4b...16657ec
